File: core/core-api/src/main/java/com/coffee/gu/cart/CartItemManager.java
package com.coffee.gu.cart;


import com.coffee.gu.Principal;
import jakarta.transaction.Transactional;
import org.springframework.stereotype.Component;

import java.util.Optional;


@Component
public class CartItemManager {

    private final CartItemRepository cartItemRepository;

    public CartItemManager(CartItemRepository cartItemRepository) {
        this.cartItemRepository = cartItemRepository;
    }

    @Transactional
    public Long addCartItem(Principal principal, AddCartItem addCartItem, Optional<CartItem> cartItem) {
        return cartItem.map(item -> {
            if (item.getIsDeleted()) {
                item.active();
                item.applyQuantity(addCartItem.quantity());
            } else {
                item.applyQuantity(item.getQuantity() + addCartItem.quantity());
            }
            return cartItemRepository.save(item, principal).getId();
        }).orElseGet(() -> cartItemRepository.save(CartItem.create(addCartItem.menuId(), addCartItem.quantity()), principal).getId());
    }

    public void modifyCartItem(CartItem cartItem, Long quantity, Principal principal) {
        cartItem.applyQuantity(quantity);
        cartItemRepository.save(cartItem, principal);
    }

    @Transactional
    public void deleteCartItem(CartItem cartItem, Principal principal) {
        cartItem.delete();
        cartItemRepository.save(cartItem, principal);
    }
}

File: core/core-api/src/main/kotlin/com/coffee/gu/cart/CartItemManager.kt
package com.coffee.gu.cart

import com.coffee.gu.Principal
import com.coffee.gu.menu.MenuRepository
import jakarta.transaction.Transactional
import org.springframework.stereotype.Component

@Component
class CartItemManager(
    private val cartItemRepository: CartItemRepository,
    private val menuRepository: MenuRepository,
) {
    @Transactional
    fun addCartItem(principal: Principal, addCartItem: AddCartItem, cartItem: CartItem?): Long {
        val targetItem = cartItem?.apply {
            if (isDeleted) {
                active()
                applyQuantity(addCartItem.quantity)
            } else {
                applyQuantity(quantity + addCartItem.quantity)
            }
        } ?: run {
            val menu = menuRepository.findById(addCartItem.menuId)
            CartItem(menu=menu, quantity=addCartItem.quantity)
        }
        return cartItemRepository.save(targetItem, principal).id
    }

    fun modifyCartItem(cartItem: CartItem, quantity: Long, principal: Principal) {
        cartItem.applyQuantity(quantity)
        cartItemRepository.save(cartItem, principal)
    }

    @Transactional
    fun deleteCartItem(cartItem: CartItem, principal: Principal) {
        cartItem.delete()
        cartItemRepository.save(cartItem, principal)
    }
}
